tools/base/libboard/board_vc/board_vc.c

```diff
@@ -0,0 +1,749 @@
+{
+	fpga_result res                 = FPGA_OK;
+	fpga_result resval              = FPGA_OK;
+	unsigned char buf[8]            = {0};
+	fpga_object mac_object;
+
+	if (mac_info == NULL) {
+		FPGA_ERR("Invalid Input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	res = fpgaTokenGetObject(token, SYSFS_EEPROM, &mac_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get token object");
+		return res;
+	}
+
+	res = fpgaObjectRead(mac_object, (uint8_t *)buf, 0, sizeof(buf), 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to read object ");
+		resval = res;
+		goto out_destroy_mac;
+	}
+
+	memcpy_s(mac_info, len, buf, sizeof(buf));
+
+out_destroy_mac:
+	res = fpgaDestroyObject(&mac_object);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to destroy object");
+	}
+
+	return resval;
+}
+
+// Read pkvl versoin
+fpga_result print_pkvl_version(fpga_token token)
+{
+	fpga_result res                 = FPGA_OK;
+	fpga_result resval              = FPGA_OK;
+	char ver_a_buf[VER_BUF_SIZE]    = { 0 };
+	char ver_b_buf[VER_BUF_SIZE]    = { 0 };
+	uint32_t size                   = 0;
+	fpga_object pkvl_a_object;
+	fpga_object pkvl_b_object;
+
+
+	res = fpgaTokenGetObject(token, SYSFS_PKVL_A_VER, &pkvl_a_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to get token object");
+		return res;
+	}
+
+	res = fpgaTokenGetObject(token, SYSFS_PKVL_B_VER, &pkvl_b_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to get token object");
+		resval = res;
+		goto out_destroy_obj_a;
+	}
+
+
+	res = fpgaObjectGetSize(pkvl_a_object, &size, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get object size");
+		resval = res;
+		goto out_destroy_obj_b;
+	}
+
+	res = fpgaObjectRead(pkvl_a_object, (uint8_t *)ver_a_buf, 0, size, 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to read object ");
+		resval = res;
+		goto out_destroy_obj_b;
+	}
+
+	res = fpgaObjectGetSize(pkvl_b_object, &size, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get object size");
+		resval = res;
+		goto out_destroy_obj_b;
+	}
+
+	res = fpgaObjectRead(pkvl_b_object, (uint8_t *)ver_b_buf, 0, size, 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to read object ");
+		resval = res;
+		goto out_destroy_obj_b;
+	}
+
+	printf("%-32s : %s", "Retimer A Version", ver_a_buf);
+	printf("%-32s : %s", "Retimer B Version", ver_b_buf);
+
+out_destroy_obj_b:
+	res = fpgaDestroyObject(&pkvl_b_object);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to destroy object");
+	}
+
+out_destroy_obj_a:
+	res = fpgaDestroyObject(&pkvl_a_object);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to destroy object");
+	}
+
+	return resval;
+}
+
+// print mac information
+fpga_result print_mac_info(fpga_token token)
+{
+	fpga_result res               = FPGA_OK;
+	unsigned char buf[8]          = { 0 };
+	int i                         = 0;
+	int n                         = 0;
+	pkvl_mac mac;
+
+	res = read_mac_info(token, buf, SYSFS_MAX_SIZE);
```

Comment:
We call `read_mac_info` with buf being 8 bytes long but the third argument is `SYSFS_MAX_SIZE` which is way bigger than 8 bytes.

Inside the implementation of `read_mac_info`, on line 501, we call memcpy_s like so:

```C
	memcpy_s(mac_info, len, buf, sizeof(buf));
```

Where `mac_info` is `buf` and `len` is `SYSFS_MAX_SIZE`. If `len` is used to limit how many bytes can be written to the buffer, `mac_info`, then we can potentially overrun this buffer.

Reply:
Fixed